core/embed/rust/src/ui/model_tt/component/progress.rs

```diff
@@ -0,0 +1,136 @@
+use core::mem;
+
+use crate::ui::{
+    component::{
+        base::ComponentExt,
+        paginated::Paginate,
+        text::paragraphs::{Paragraph, Paragraphs},
+        Child, Component, Event, EventCtx, Label, Never, Pad,
+    },
+    display::{self, Font},
+    geometry::{Insets, Rect},
+    model_tt::constant,
+    util::animation_disabled,
+};
+
+use super::theme;
+
+pub struct Progress<T, F>
+where
+    F: Fn(&str) -> T,
+{
+    title: Child<Label<T>>,
+    value: u16,
+    loader_y_offset: i16,
+    indeterminate: bool,
+    description: Child<Paragraphs<Paragraph<T>>>,
+    description_pad: Pad,
+    description_func: F,
+}
+
+impl<T, F> Progress<T, F>
+where
+    T: AsRef<str>,
+    F: Fn(&str) -> T,
+{
+    const AREA: Rect = constant::screen().inset(theme::borders());
+
+    pub fn new(title: T, indeterminate: bool, description: T, description_func: F) -> Self {
+        Self {
+            title: Label::centered(title, theme::label_progress()).into_child(),
+            value: 0,
+            loader_y_offset: 0,
+            indeterminate,
+            description: Paragraphs::new(
+                Paragraph::new(&theme::TEXT_NORMAL, description).centered(),
+            )
+            .into_child(),
+            description_pad: Pad::with_background(theme::BG),
+            description_func,
+        }
+    }
+}
+
+impl<T, F> Component for Progress<T, F>
+where
+    T: AsRef<str>,
+    F: Fn(&str) -> T,
+{
+    type Msg = Never;
+
+    fn place(&mut self, _bounds: Rect) -> Rect {
+        let description_lines = 1 + self
+            .description
+            .inner()
+            .inner()
+            .content()
+            .as_ref()
+            .chars()
+            .filter(|c| *c == '\n')
+            .count() as i16;
+        let (title, rest) = Self::AREA.split_top(self.title.inner().size().y);
+        let (loader, description) =
+            rest.split_bottom(Font::NORMAL.line_height() * description_lines);
+        let loader = loader.inset(Insets::top(theme::CONTENT_BORDER));
+        self.title.place(title);
+        self.loader_y_offset = loader.center().y - constant::screen().center().y;
+        self.description.place(description);
+        self.description_pad.place(description);
+        Self::AREA
+    }
+
+    fn event(&mut self, ctx: &mut EventCtx, event: Event) -> Option<Self::Msg> {
+        if let Event::Progress(new_value, new_description) = event {
+            if mem::replace(&mut self.value, new_value) != new_value {
+                if !animation_disabled() {
+                    ctx.request_paint();
+                }
+                self.description.mutate(ctx, |ctx, para| {
+                    if para.inner_mut().content().as_ref() != new_description {
+                        para.inner_mut()
+                            .update((self.description_func)(new_description));
```

Comment:
this is very weird. how about instead we drop `description_func` altogether and give `Paragraph<StrBuffer>` a method `update_with(&mut self, &str)` (and later if we need it, we can add one for `Paragraph<&str>` too)

Reply:
ooohh but that would only work if we specified a trait for it, otherwise it won't build for `Progress<T>`, only for `Progress<StrBuffer>`

Reply:
We can add something like `fn from_str(&str) -> T` to the `ParagraphStrType` trait, or perhaps make it a supertrait of `From<&str>`/`TryFrom<&str>`, it's not too exotic a feature to require from a string type.

Reply:
leaving that up to you.
i think the fully correct solution here is "use StrBuffer for everything", moving `StrBuffer` to crate root and making it not depend on micropython, so I'd like to try that in any case (the main blocker being any possible adverse effect on bootloader code)